ONSdigital/rm-collection-exercise-service
# Spring boot update

# What and why?
As suggested earlier spring boot 2.5.5 has breaking change in the latest spring boot version is on the datasource initialisation. There are a lot of constraint to keep moving with spring boot versions.  it seems that liquibase is executed before Hibernate is initialised as per the new spring boot structure.

As the system is already in prod, there is not a good way to update liquibase with schema creation. We will have to rely on build upon the schema creation performed by hibernate. which requires `spring.jpa.defer-datasource-initialization` to be set to true with the new springboot structure. Now this will conflict with the default liquibase initialisation as there will be a circular dependence with the entity manager.

Keeping all the constraint in mind this is the solution I have come up with. I don't like this solution myself as its seems a workaround for the constraint created due to bad coding practices. Although I can't think of any other solution. If any one has any other ideas that would be great.

# How to test?

- Deploy fresh copy of your environment
- Delete the `collectionexercise` schema
- Deploy this PR to your environment, it'll error but the tables will be created as expected and it should start up fine
- Run acceptance tests, they should pass
- `kc rollout restart deployment collection-exercise` to restart it, the data should remain and it shouldn't error on startup anymore.

# Trello
https://trello.com/c/rZkpcZB7
